shell/platform/fuchsia/flutter/engine.cc

```diff
@@ -920,10 +915,25 @@ void Engine::WarmupSkps(
 
       // Tell raster task runner to warmup have the compositor
       // context warm up the newly deserialized picture
-      raster_task_runner->PostTask([skp_warmup_surface, picture,
-                                    surface_producer, completion_callback, i,
+      raster_task_runner->PostTask([picture, size, surface_producer,
+                                    completion_callback, i,
                                     count = skp_mappings.size()] {
         TRACE_DURATION("flutter", "WarmupSkp");
+
+        // We use a raw pointer here because we want to keep this alive until
+        // all gpu work is done and the callbacks skia takes for this are
+        // function pointers so we are unable to use a lambda that captures the
+        // smart pointer.
+        static SurfaceProducerSurface* skp_warmup_surface =
```

Comment:
I was confused about the static inside a lambda here, so I looked it up:  https://stackoverflow.com/questions/8391058/how-do-static-variables-in-lambda-function-objects-work/35038341 and https://stackoverflow.com/questions/55510/when-do-function-level-static-variables-get-allocated-initialized/58804#58804

This will work for the first time shader warmup is done within a given process, but subsequent warmup requests will fail mysteriously with the "Failed to create" error.  This is because the static will be re-used across all lambda invocations....

Since you delete skp_warmup_surface when you're done with it below, it will be nullptr forever after and subsequent shader warmups will always fail (and it'll be really hard to debug).

It seems like you'll have to post 1 task to create the surface, and then post a bunch more inside the for loop which use the surface as a capture.

Reply:
oh damn good catch I didnt even think about that with the lambda, lemme chew on that and I'll cook up a solution. I guess i can move the pointer declaration to the outer function and then just initialize the pointer in the lambda. That way well get a new pointer for each invocation of WarmupSkps